File: desktop/launcher/PortSelector.cs
using System.Net;
using System.Net.Sockets;

namespace LogExplorer.Launcher;

/// <summary>
/// Legacy Remediation Slice 9 §B/§C: 3434 is preferred but never assumed
/// free. This is only ever called after <see cref="SingleInstanceGuard"/>
/// has already confirmed this is the one true Log Explorer instance, so a
/// bind failure on the preferred port here means an unrelated process
/// owns it, not another Log Explorer - no port-based identity check is
/// layered on top of the mutex-based one in <see cref="SingleInstanceGuard"/>.
/// </summary>
internal static class PortSelector
{
    public const int PreferredPort = 3434;

    /// <summary>
    /// Returns 3434 if free; otherwise an OS-assigned free loopback port
    /// (never a predictable scan across a range - the mission's own
    /// explicit "do not create an unsafe predictable scan" instruction).
    /// There is an inherent, small TOCTOU race between this probe and the
    /// backend actually binding the port a moment later - the same
    /// unavoidable race any "ask the OS for a free port" approach has
    /// (including Spring Boot's own <c>server.port=0</c>) - acceptable
    /// here since a bind failure on startup fails loudly rather than
    /// silently.
    /// </summary>
    public static int SelectPort()
    {
        if (TryBindLoopback(PreferredPort))
        {
            return PreferredPort;
        }

        using var listener = new TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }

    private static bool TryBindLoopback(int port)
    {
        try
        {
            using var listener = new TcpListener(IPAddress.Loopback, port);
            listener.Start();
            listener.Stop();
            return true;
        }
        catch (SocketException)
        {
            return false;
        }
    }
}

File: desktop/launcher-macos/src/main/java/com/logexplorer/desktoplauncher/PortSelector.java
package com.logexplorer.desktoplauncher;

import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;

/**
 * REL-1 (mirrors desktop/launcher/PortSelector.cs): 3434 is preferred but
 * never assumed free. Falls back to an OS-assigned free loopback port -
 * never a predictable scan across a range.
 */
final class PortSelector {

    static final int PREFERRED_PORT = 3434;

    private PortSelector() {
    }

    static int selectPort() {
        if (tryBindLoopback(PREFERRED_PORT)) {
            return PREFERRED_PORT;
        }
        try (ServerSocket socket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress())) {
            return socket.getLocalPort();
        } catch (IOException e) {
            throw new IllegalStateException("Could not obtain any free loopback port from the OS.", e);
        }
    }

    private static boolean tryBindLoopback(int port) {
        try (ServerSocket socket = new ServerSocket(port, 0, InetAddress.getLoopbackAddress())) {
            return true;
        } catch (IOException e) {
            return false;
        }
    }
}
